extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeJavascriptExtractor.java

```diff
@@ -0,0 +1,101 @@
+package org.schabi.newpipe.extractor.services.youtube;
+
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+import org.schabi.newpipe.extractor.NewPipe;
+import org.schabi.newpipe.extractor.exceptions.ParsingException;
+import org.schabi.newpipe.extractor.localization.Localization;
+import org.schabi.newpipe.extractor.utils.Parser;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Youtube restricts streaming their media in multiple ways by requiring clients to apply a cipher function
+ * on parameters of requests.
+ * The cipher function is sent alongside as a JavaScript function.
+ * <p>
+ * This class handling fetching the JavaScript file in order to allow other classes to extract the needed functions.
+ */
+public class YoutubeJavascriptExtractor {
+
+    private static final String HTTPS = "https:";
+    private static String cachedJavascriptCode;
+
+    /**
+     * Extracts the JavaScript file. The result is cached, so subsequent calls use the result of previous calls.
+     *
+     * @param videoId Does not influence the result, but a valid video id can prevent tracking
+     * @return The whole javascript file as a string.
+     * @throws ParsingException If the extraction failed.
+     */
+    @Nonnull
+    public static String extractJavascriptCode(String videoId) throws ParsingException {
+        if (cachedJavascriptCode == null) {
+            final YoutubeJavascriptExtractor extractor = new YoutubeJavascriptExtractor();
```

Comment:
what is the benefit of creating a `YoutubeJavascriptExtractor` instance? Creating a new object takes some time and I do not see any non-static class attributes. Is there a reason to not make the private class methods static?

Reply:
This comes more out of habit than any reason. Only other noteable space where I'm developing is server-side, where the performance impact of object creation is negligible.

I'll adjust it for now until a reason for non-static occurs.